pkg/storaged/pages.jsx

```diff
@@ -0,0 +1,691 @@
+    type_extra, id_extra,
+    page_name, page_icon, page_key, page_location, page_size,
+    page_block,
+    for_summary,
+    has_warning, has_danger, job_path,
+    component, props,
+    actions,
+}) {
+    if (page_block) {
+        page_location = [block_location(page_block)];
+        page_name = block_short_name(page_block);
+        page_size = page_block.Size;
+        job_path = page_block.path;
+    }
+    return {
+        title,
+        location,
+        next,
+        type_extra,
+        page_name,
+        page_icon,
+        page_key,
+        page_location,
+        page_size,
+        for_summary,
+        component,
+        props,
+        has_warning,
+        has_danger,
+        job_path,
+        actions: actions ? actions.filter(a => !!a) : null,
+        id_extra,
+    };
+}
+
+export function register_crossref(crossref) {
+    const val = crossrefs.get(crossref.key) || [];
+    crossref.actions = crossref.actions ? crossref.actions.filter(a => !!a) : null;
+    val.push(crossref);
+    crossrefs.set(crossref.key, val);
+}
+
+export function get_crossrefs(key) {
+    return crossrefs.get(key);
+}
+
+/* Getting the page for a navigation location.
+ *
+ * We have a special "not found" page that is returned when there is
+ * no real page at the given location.
+ */
+
+const NotFoundCard = ({ card }) => {
+    return <span>{_("Not found")}</span>;
+};
+
+const not_found_page = new_page(null, new_card({ page_name: _("Not found"), component: NotFoundCard }));
+
+export function get_page_from_location(location) {
+    if (!pages)
+        return not_found_page;
+
+    return pages.get(JSON.stringify(location)) || not_found_page;
+}
+
+/* Common UI things
+ */
+
+export function navigate_away_from_card(card) {
+    if (!card)
+        return;
+
+    const loc = cockpit.location;
+    const page = card.page;
+    if (page.parent && JSON.stringify(loc.path) == JSON.stringify(page.location))
+        loc.go(page.parent.location);
+}
+
+export function navigate_to_new_card_location(card, location) {
+    if (!card)
+        return;
+
+    const loc = cockpit.location;
+    const page = card.page;
+    if (JSON.stringify(loc.path) == JSON.stringify(page.location))
+        loc.go(location);
+}
+
+function make_menu_item(action) {
+    return <StorageMenuItem key={action.title} onClick={() => action.action(true)}
+                            danger={action.danger} excuse={action.excuse}>
+        {action.title}
+    </StorageMenuItem>;
+}
+
+function make_page_kebab(page) {
+    const items = [];
+
+    function is_interesting(action) {
+        // Impossible actions are omitted from the kebab
+        // menus. The user needs to go to the page itself to see
+        // the reason for why something is impossible.
+        //
+        // Pages that don't really exist ("Free space" and
+        // "Extended partition") get all their actions in the
+        // menu, because there is no other place.
+
+        return true;
+    }
```

Comment:
Do you still intend to fill this with some actual logic, or is this just a leftover from iterating?

Reply:
Upps, leftover.